src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1378,43 +1363,64 @@ private static bool HasIdentityConversionInternal(TypeSymbol type1, TypeSymbol t
-                TypeCompareKind.AllIgnoreOptions;
+            var compareKind = TypeCompareKind.AllIgnoreOptions |
+                (includeNullability ?
+                    TypeCompareKind.CompareNullableModifiersForReferenceTypes |
+                        (unknownNullableModifierMatchesAny ? TypeCompareKind.UnknownNullableModifierMatchesAny : 0) :
+                    0);
             return type1.Equals(type2, compareKind);
         }
 
         private bool HasIdentityConversionInternal(TypeSymbol type1, TypeSymbol type2)
         {
-            return HasIdentityConversionInternal(type1, type2, IncludeNullability);
+            return HasIdentityConversionInternal(type1, type2, IncludeNullability, UnknownNullableModifierMatchesAny);
         }
 
-        internal static bool HasTopLevelNullabilityIdentityConversion(TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations destination)
+        internal bool HasTopLevelNullabilityIdentityConversion(TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations destination)
         {
+            if (!IncludeNullability)
+            {
+                return true;
+            }
             bool? sourceIsNullable = source.IsNullable;
             bool? destinationIsNullable = destination.IsNullable;
-            return sourceIsNullable == null || destinationIsNullable == null || sourceIsNullable == destinationIsNullable;
-        }
-
-        internal static bool HasTopLevelNullabilityImplicitConversion(TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations destination)
-        {
-            return HasTopLevelNullabilityImplicitConversion(source.IsNullable, destination.IsNullable);
+            if (sourceIsNullable == destinationIsNullable)
+            {
+                return true;
+            }
+            return UnknownNullableModifierMatchesAny &&
+                (sourceIsNullable == null || destinationIsNullable == null);
         }
 
-        internal static bool HasTopLevelNullabilityImplicitConversion(bool? sourceIsNullable, bool? destinationIsNullable)
+        internal bool HasTopLevelNullabilityImplicitConversion(TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations destination)
         {
-            return sourceIsNullable != true || destinationIsNullable != false;
+            if (!IncludeNullability)
+            {
+                return true;
+            }
+            bool? sourceIsNullable = source.IsNullable;
+            bool? destinationIsNullable = destination.IsNullable;
+            if (sourceIsNullable == destinationIsNullable)
+            {
+                return true;
+            }
+            switch (sourceIsNullable)
+            {
+                case true:
+                    return UnknownNullableModifierMatchesAny && destinationIsNullable == null;
+                case false:
+                    return true;
+                default:
```

Comment:
>default [](start = 16, length = 7)

case null?

Reply:
Control flow analysis reports an error.

---
In reply to: [221330550](https://github.com/dotnet/roslyn/pull/29727#discussion_r221330550) [](ancestors = 221330550)